Sources/BSON/Document.swift

```diff
@@ -14,10 +39,44 @@ public struct Document {
 
     internal init(_ elements: [BSON]) { fatalError("Unimplemented") }
 
-    internal init(keyValuePairs: [(String, BSON)]) { fatalError("Unimplemented") }
+    internal init(keyValuePairs: [(String, BSON)]) {
+        self.keySet = Set(keyValuePairs.map { $0.0 })
+        guard self.keySet.count == keyValuePairs.count else {
+            fatalError("Dictionary \(keyValuePairs) contains duplicate keys")
+        }
+
+        self._buffer = BSON_ALLOCATOR.buffer(capacity: 100)
+
+        let start = self._buffer.writerIndex
+
+        // reserve space for our size that will be calculated
+        self._buffer.writeInteger(0, endianness: .little, as: Int32.self)
+
+        for (key, bson) in keyValuePairs {
+            self._buffer.writeInteger(UInt8(bson.bsonValue.bsonType.rawValue), as: UInt8.self)
+            self._buffer.writeCString(key)
+            // swiftlint:disable:next force_try
+            try! bson.bsonValue.write(to: &self._buffer)
```

Comment:
I guess we are `try!`ing here since `BSONValue.write` throws. however, it looks like so far neither of our implementations actually need to throw, since `writeInteger` doesn't. maybe we can change the protocol method to not throw, and then we don't need to `try!` here?